File: shared/schemas/telemetry.py
"""
Telemetry Schema - Shared data structures for game telemetry.

This is the SOURCE OF TRUTH for telemetry data structures.
TypeScript and C++ versions should be kept in sync with this file.
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Any

import pendulum
from pendulum import DateTime


class BiomeId(str, Enum):
    """Valid biome identifiers."""
    ABYSS = "abyss"
    SCORCH = "scorch"
    RUINS = "ruins"
    AQUA = "aqua"
    BOTANY = "botany"
    THEATER = "theater"
    EXODUS = "exodus"
    BRINK = "brink"
    VECTOR = "vector"
    UPRISING = "uprising"


class ActionType(str, Enum):
    """Types of player actions."""
    MOVEMENT = "movement"
    INTERACTION = "interaction"
    TOOL_USE = "tool_use"
    COMMUNICATION = "communication"
    COMBAT = "combat"
    CONSTRUCTION = "construction"
    PROBLEM_SOLVE = "problem_solve"


@dataclass
class Vector3:
    """3D vector for positions and directions."""
    x: float
    y: float
    z: float


@dataclass
class PlayerInput:
    """Raw player input data."""
    timestamp_ms: int
    input_type: str
    value: float
    device: str  # "keyboard", "mouse", "controller"


@dataclass
class TelemetryEvent:
    """A single telemetry event from the game client."""
    event_id: str
    session_id: str
    player_id_hash: str  # Anonymized player identifier
    biome_id: BiomeId
    timestamp: DateTime
    action_type: ActionType
    position: Vector3
    velocity: Vector3
    inputs: list[PlayerInput]
    context: dict[str, Any] = field(default_factory=dict)


@dataclass
class SessionStart:
    """Session start event."""
    session_id: str
    player_id_hash: str
    biome_id: BiomeId
    started_at: DateTime
    client_version: str
    hardware_hash: str  # Anonymized hardware fingerprint


@dataclass
class SessionEnd:
    """Session end event."""
    session_id: str
    ended_at: DateTime
    reason: str  # "normal", "disconnect", "timeout", "kicked"
    duration_seconds: int
    events_count: int


@dataclass
class SolutionAttempt:
    """A player's attempt to solve a problem."""
    attempt_id: str
    session_id: str
    player_id_hash: str
    biome_id: BiomeId
    problem_id: str
    started_at: DateTime
    completed_at: DateTime | None
    success: bool
    solution_data: dict[str, Any]
    tools_used: list[str]
    collaborators: list[str]  # Other player hashes if collaborative

File: shared/schemas/telemetry.ts
/**
 * Telemetry Schema - TypeScript version
 *
 * SYNC WITH: shared/schemas/telemetry.py (source of truth)
 *
 * This file should be kept in sync with the Python version.
 * Any changes to the Python schema should be reflected here.
 */

/**
 * Valid biome identifiers.
 */
export enum BiomeId {
  ABYSS = "abyss",
  SCORCH = "scorch",
  RUINS = "ruins",
  AQUA = "aqua",
  BOTANY = "botany",
  THEATER = "theater",
  EXODUS = "exodus",
  BRINK = "brink",
  VECTOR = "vector",
  UPRISING = "uprising",
}

/**
 * Types of player actions.
 */
export enum ActionType {
  MOVEMENT = "movement",
  INTERACTION = "interaction",
  TOOL_USE = "tool_use",
  COMMUNICATION = "communication",
  COMBAT = "combat",
  CONSTRUCTION = "construction",
  PROBLEM_SOLVE = "problem_solve",
}

/**
 * 3D vector for positions and directions.
 */
export interface Vector3 {
  x: number;
  y: number;
  z: number;
}

/**
 * Raw player input data.
 */
export interface PlayerInput {
  timestampMs: number;
  inputType: string;
  value: number;
  device: "keyboard" | "mouse" | "controller";
}

/**
 * A single telemetry event from the game client.
 */
export interface TelemetryEvent {
  eventId: string;
  sessionId: string;
  playerIdHash: string;
  biomeId: BiomeId;
  timestamp: string; // ISO 8601 datetime
  actionType: ActionType;
  position: Vector3;
  velocity: Vector3;
  inputs: PlayerInput[];
  context?: Record<string, unknown>;
}

/**
 * Session start event.
 */
export interface SessionStart {
  sessionId: string;
  playerIdHash: string;
  biomeId: BiomeId;
  startedAt: string; // ISO 8601 datetime
  clientVersion: string;
  hardwareHash: string;
}

/**
 * Session end event.
 */
export interface SessionEnd {
  sessionId: string;
  endedAt: string; // ISO 8601 datetime
  reason: "normal" | "disconnect" | "timeout" | "kicked";
  durationSeconds: number;
  eventsCount: number;
}

/**
 * A player's attempt to solve a problem.
 */
export interface SolutionAttempt {
  attemptId: string;
  sessionId: string;
  playerIdHash: string;
  biomeId: BiomeId;
  problemId: string;
  startedAt: string; // ISO 8601 datetime
  completedAt: string | null; // ISO 8601 datetime or null
  success: boolean;
  solutionData: Record<string, unknown>;
  toolsUsed: string[];
  collaborators: string[];
}
